lib/rules/no-extra-parens.js

```diff
@@ -337,16 +337,44 @@ module.exports = {
         }
 
         /**
+         * Check if a member expression contains a call expression
+         * @param {ASTNode} node MemberExpression node to evaluate
+         * @returns {boolean} true if found, false if not
+         */
+        function doesMemberExpressionContainCallExpression(node) {
+            let currentNode = node.object;
+            let currentNodeType = node.object.type;
+
+            while (currentNodeType !== "Identifier" && currentNodeType !== "CallExpression") {
+                currentNode = currentNode.object;
```

Comment:
This function will sometimes throw an error, because a `MemberExpression` node doesn't always start with an `Identifier` or a `CallExpression`. For example, this implementation crashes when linting the following code:

```js
({}.foo)()
```

with the following error:

```
     TypeError: Cannot read property 'type' of undefined
      at doesMemberExpressionContainCallExpression (lib/rules/no-extra-parens.js:350:46)
      at checkCallNew (lib/rules/no-extra-parens.js:375:25)
      at EventEmitter.CallExpression (lib/rules/no-extra-parens.js:530:17)
      at NodeEventGenerator.applySelector (lib/util/node-event-generator.js:265:26)
      at NodeEventGenerator.applySelectors (lib/util/node-event-generator.js:294:22)
      at NodeEventGenerator.enterNode (lib/util/node-event-generator.js:308:14)
      at CodePathAnalyzer.enterNode (lib/code-path-analysis/code-path-analyzer.js:606:23)
      at Traverser.enter (lib/linter.js:924:32)
      at Traverser.__execute (node_modules/estraverse/estraverse.js:397:31)
      at Traverser.traverse (node_modules/estraverse/estraverse.js:501:28)
      at Traverser.traverse (lib/util/traverser.js:31:22)
      at Linter.verify (lib/linter.js:921:24)
      at runRuleForItem (lib/testers/rule-tester.js:368:38)
      at testValidTemplate (lib/testers/rule-tester.js:400:28)
      at Context.RuleTester.it (lib/testers/rule-tester.js:532:25)
```

To fix this, I think you should update the check in `while (currentNodeType !== "Identifier" && currentNodeType !== "CallExpression")` to simply be `while (currentNode.type === "MemberExpression")`. That way, you can be guaranteed that `currentNode.object` always exists, so you won't end up with errors like this.